File: LeetCode/c++/GrayCode.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<int> grayCode(int n) 
{
	vector<int> v;
	v.push_back(0);

	if (n == 0)
		return v;

	v.push_back(1);

	for (int i = 1; i < n; i++)
	{
		//计算要加的2次幂数
		int addition = 1;
		int cnt = 0;
		while (cnt < i)
		{
			addition *= 2;
			cnt++;
		}

		//从容器开始计算格雷码
		int nextGrayCode;
		for (int j = addition - 1; j >= 0; j--) //从2次幂到0加上addition
		{
			nextGrayCode = v[j] + addition;
			v.push_back(nextGrayCode);
		}
	}

	return v;
}

int main()
{
	vector<int> v;
	v = grayCode(0);
	for (int i = 0; i < v.size(); i++)
		cout << v[i] << " ";

	return 0;
}

File: LeetCode/iter1/python/GrayCode.py
class Solution(object):
    def grayCode(self, n):
        """
        :type n: int
        :rtype: List[int]
        """
        result = [0]
        for i in range(n):
            #得到最后一位的差距
            addition = 1
            cnt = 0
            while(cnt<i):
                addition *= 2
                cnt += 1
			
            for j in range(addition-1,-1,-1):
                result.append(result[j]+addition)
			
        return result
		

#测试程序
s = Solution()
for i in range(5):
	print(s.grayCode(i))